RFC-0001-prsestissimo-metrics.md

```diff
@@ -0,0 +1,251 @@
+1. PrestoServer registers PrestoCPP counters and Velox runtime metrics at the launch of the application.
+   Note that we must set the static flag in `BaseStatsReporter::registered=true` before calling register.
+   **_PrestoServer::run()→registerPrestoCppCounters()→DEFINE_METRIC(<metric_name>, metric_type)_**
+2. PrestoServer starts **_PeriodicTaskManager_** which registers schedulers (callbacks) to collect metrics at regular intervals and report it through BaseStatsReporter.
+   ** _PrestoServer::run () → PeriodicTaskManager::start () →RECORD_METRIC_VALUE (<metric_name>, <value>)_ **
+3. When RECORD_METRIC_VALUE is invoked for a metric, the implementer of BaseStatsReporter must ensure the value is appropriately recorded depending on its StatType.
+
+### Design to export runtime metrics:
+
+#### How to capture metrics BaseStatsReporter Interface
+
+BaseStatsReporter has following APIS that must be implemented for user to capture metrics:
+
+`void registerMetricExportType(<key>, <type>)` to Register a metric of type COUNT, SUM, AVG.
+
+`void registerHistogramMetricExportType(<key>, <bucketwidth>,<min>,<max>, <vector<pct>>)`  to register a HISTOGRAM.
+
+`void addMetricValue(<key>, <value>)` to add a value for a metric key previously registered.
+
+`void addHistogramMetricValue(<key, <value>)` to add a value to a histogram type metric key.
+
+The metrics are reported by one of the [above] reporting flows. The user implementing BaseStatsReporter must adhere to the above Metrics Semantics of Velox while storing the metrics. That implies, the COUNT type metric must not be overridden and continuously grow with time,  On the other hand, the sum metrics which reports the change in aggregated stats over a period of time will overwrite the value against that metric key. Same applies to AVG, RATE and HISTOGRAMs.
+
+At any instant, we have a snapshot of metrics maintained which is overwritten on new updates.
+
+#### How to Export:
+**Prestissimo as exporter:**
+To keep it simple, the prestissimo worker will behave like an exporter and exposes the REST endpoint `/v1/info/health/metrics`. The Prometheus server can be configured to pull metrics from the worker itself. Again, it is up to the user if they want to introduce another layer in between the Time series DB and the worker.
+
+**A sidecar metric exporter:**
+A sidecar HTTP server that fetches metrics from Prestissimo. This would require launching as a child process of Prestissimo or as a separate container in the same POD as Prestissimo.
+Pros: In case of Prestissimo crash, we still have a way to get the last set of metrics from the node.
+      To support graceful failure handling, we may have to dump metrics to disk and the exporter must read from disk when Prestissimo is down.
+Cons: Spawning a new process as a sidecar would take additional memory and CPU.
+
+Currently, we have only prototyped Prestissimo as exporter. Both approaches can be built using the prometheus-cpp library.
```

Comment:
Is the RPC communication between the presto server and metrics sidecar clear? 
I am not sure if HTTP can be used and if that itself would be an overhead. JMX  exporter uses some custom rpc and attaches to the jvm. 
If we can build a scalable sidecar, it could be a separate opensource project in its own like the jmx exporter